### PR TITLE
Add a few json functions

### PR DESCRIPTION
**Pull request by anirbanmu**
#### JSON functions implemented:
- `JSON_VALID`
- `JSON_QUOTE`
- `JSON_UNQUOTE`
- `JSON_EXTRACT`
- `JSON_REPLACE`

#### Regular functions implemented:
- `REPLACE`

#### Known issues:
- ~~Constructing JSON values using output of operators/functions that produce boolean values does not retain true/false & instead ends up using 1/0.~~
- ~~Using a string value output of JSON_EXTRACT to construct other JSON (like via JSON_REPLACE) ends up double quoting string in the final JSON. This can be worked around by doing a JSON_UNQUOTE on the output of JSON_EXTRACT before passing to JSON_REPLACE.~~
- I think using keys (for paths) that have a space in them is also not supported.

### CODE REVIEW

File: src/Expressions/Expression.php [hunk 1]
```diff
@@ -45,10 +50,33 @@ public function negate(): void {
    */
   public abstract function isWellFormed(): bool;
 
+
+  // This is not the method to override by the concrete Expression subclasses ()
+  public function evaluate(
```
Comment (ssandler) [inline src/Expressions/Expression.php:55]: should this function be final?

File: src/Parser/SQLParser.php [hunk 2]
```diff
@@ -268,13 +270,13 @@ private static function buildTokenListFromLexemes(vec<string> $tokens): token_li
         );
       } else if ($token_upper === 'TRUE') {
         $out[] = shape(
-          'type' => TokenType::NUMERIC_CONSTANT,
+          'type' => TokenType::BOOLEAN_CONSTANT,
           'value' => '1',
```
Comment (ssandler) [inline src/Parser/SQLParser.php:274]: should these values be updated to true/false?

File: src/JSONPath/JSONObject.hack [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+        return $jsonObject;
+    }
+
+    private static function matchArrayIndex(string $jsonPath): ?shape('index' => string, 'rest' => string) {
+        if (Str\is_empty($jsonPath) || $jsonPath[0] != self::TOK_SELECTOR_BEGIN) {
+            return null;
+        }
+
+        $matches = null;
+        if (\preg_match_with_matches(self::RE_ARRAY_INDEX, $jsonPath, inout $matches)) {
```
Comment (ssandler) [inline src/JSONPath/JSONObject.hack:246]: would it work to use HSL Regex here? (`HH\Lib\Regex\every_match()` I think?) We try to use HSL regex everywhere if possible - the only thing it doesn't support is dynamic regex strings but for these static regexes it should work and it provides a shape describing the match output including capture groups which helps keep things nice and safe (I also find it's nice to name the capture groups to make that easier to understand).

File: src/JSONPath/JSONObject.hack [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+            }
+
+            return $out;
+        }
+
+        return $jsonObject;
+    }
+
+    private static function matchArrayIndex(string $jsonPath): ?shape('index' => string, 'rest' => string) {
+        if (Str\is_empty($jsonPath) || $jsonPath[0] != self::TOK_SELECTOR_BEGIN) {
```
Comment (ssandler) [inline src/JSONPath/JSONObject.hack:241]: ok to use `!==` here?

**Review by ssandler — COMMENTED** [review]
Looks really good overall! Some small questions.

File: src/Expressions/Expression.php (hunk 1 above)
Comment (anirbanmu) [reply src/Expressions/Expression.php:55]: yes, fixed.

File: src/Parser/SQLParser.php (hunk 2 above)
Comment (anirbanmu) [reply src/Parser/SQLParser.php:274]: we could do that, but it's handled fine here - https://github.com/slackhq/hack-sql-fake/pull/55/files#diff-607397c4446b8000f684bf522a8a0e6561187a5fd7a6ffe3a211acfcbcbf33b9R28

let me know if you'd prefer we store the 'TRUE', 'FALSE' here then convert to a boolean in https://github.com/slackhq/hack-sql-fake/pull/55/files#diff-607397c4446b8000f684bf522a8a0e6561187a5fd7a6ffe3a211acfcbcbf33b9R28 using `$token['value'] === 'TRUE' ? true : false`

File: src/JSONPath/JSONObject.hack (hunk 4 above)
Comment (anirbanmu) [reply src/JSONPath/JSONObject.hack:241]: yep, for sure. i probably just forgot that php/hack has != and !== for a second... fixed.

File: src/JSONPath/JSONObject.hack (hunk 3 above)
Comment (anirbanmu) [reply src/JSONPath/JSONObject.hack]: Switched them over. Also learned in the process, that you can't use a `re"\\"` prefix string as a constant...

**Review by ssandler — APPROVED** [review]

**Pull request merged**